service/src/main/java/bio/terra/workspace/service/iam/ControlledResourceSamPolicyBuilder.java

```diff
@@ -0,0 +1,206 @@
+        }
+        // Application is always editor on its resources; other policies are inherited
+        AccessPolicyMembershipV2 editorPolicy =
+            new AccessPolicyMembershipV2()
+                .addRolesItem(ControlledResourceIamRole.EDITOR.toSamRole());
+        addApplicationResourceEditorPolicy(editorPolicy, userRequest);
+        request.putPoliciesItem(ControlledResourceIamRole.EDITOR.toSamRole(), editorPolicy);
+        break;
+
+      case APPLICATION_PRIVATE:
+        policyMap = makePolicyMap();
+        // Application is always editor
+        addApplicationResourceEditorPolicy(
+            policyMap.get(ControlledResourceIamRole.EDITOR), userRequest);
+        // if we have an assigned user, set up their permission
+        if (privateUser != null) {
+          policyMap.get(privateIamRole).addMemberEmailsItem(privateUser);
+        }
+        applyPolicyMap(request, policyMap);
+        break;
+    }
+  }
+
+  /**
+   * For the private resources we always fill in READER, WRITER, and EDITOR even if empty. This
+   * generates a map by role of the policy.
+   */
+  private Map<ControlledResourceIamRole, AccessPolicyMembershipV2> makePolicyMap() {
+    Map<ControlledResourceIamRole, AccessPolicyMembershipV2> policyMap = new HashMap<>();
+    policyMap.put(
+        ControlledResourceIamRole.READER,
+        new AccessPolicyMembershipV2().addRolesItem(ControlledResourceIamRole.READER.toSamRole()));
+    policyMap.put(
+        ControlledResourceIamRole.WRITER,
+        new AccessPolicyMembershipV2().addRolesItem(ControlledResourceIamRole.WRITER.toSamRole()));
+    policyMap.put(
+        ControlledResourceIamRole.EDITOR,
+        new AccessPolicyMembershipV2().addRolesItem(ControlledResourceIamRole.EDITOR.toSamRole()));
+    return policyMap;
+  }
+
+  /**
+   * Given the map created in @{@link #makePolicyMap()}, populate the resource request with the
+   * policies.
+   *
+   * @param request resource request for setting the roles on the Sam resource creation
+   * @param policyMap previously created policy map
+   */
+  private void applyPolicyMap(
+      CreateResourceRequestV2 request,
+      Map<ControlledResourceIamRole, AccessPolicyMembershipV2> policyMap) {
+    for (Entry<ControlledResourceIamRole, AccessPolicyMembershipV2> entry : policyMap.entrySet()) {
+      request.putPoliciesItem(entry.getKey().toSamRole(), entry.getValue());
+    }
+  }
+
+  /**
+   * Add WSM's service account as the owner of a controlled resource in Sam. Used for admin
+   * reassignment of resources. This assumes samService.initialize() has already been called, which
+   * should happen on start.
+   */
+  private void addWsmResourceOwnerPolicy(CreateResourceRequestV2 request)
+      throws InterruptedException {
+    try {
+      AuthenticatedUserRequest wsmRequest =
+          new AuthenticatedUserRequest().token(Optional.of(samService.getWsmServiceAccountToken()));
+      String wsmSaEmail = samService.getUserEmailFromSam(wsmRequest);
```

Comment:
I don't think I changed this logic; just moved it around.
I'm not understanding your suggestion to change it. What is ADC?

Reply:
Application Default Credentials I believe.

Reply:
So the fix would be to replace
```
      AuthenticatedUserRequest wsmRequest =
          new AuthenticatedUserRequest().token(Optional.of(samService.getWsmServiceAccountToken()));
      String wsmSaEmail = samService.getUserEmailFromSam(wsmRequest);
```
with
```
    String wsmSaEmail = GcpUtils.getWsmSaEmail();
```
??

Reply:
I think so but I'll defer to @zloery

Reply:
The one use I see requires the crlService to get the application credentials. That would require plumbing CrlService into the `ControlledResourceSamPolicyBuilder` :(

Reply:
Yep, ADC is Application Default Credentials, and that's the code change exactly. That would let you read the WSM SA's email directly from the SA credentials instead of having Sam introspect it. I don't think you changed it from the original in SamService, it's just an improvement I found later on and didn't add in there.

Looking at the code a little more, the one crlService method you need (`getApplicationCredentials`) really should be static, since it's just wrapping a static call to `GoogleCredentials.getApplicationDefault()`, which also solves the problem here.

Reply:
alternatively, I wonder if we should just pull the static `GoogleCredentials.getApplicationDefault()` directly into `GcpUtils.getWsmSaEmail`

Reply:
OK, I made a variant of `getWsmSaEmail` in `GcpUtils` and call it from here instead of the Sam calls.
Thanks for the help!